pkg/filter/stream/httpwasm/host.go

```diff
@@ -0,0 +1,315 @@
+	if hdr := f.reqHeaders; hdr == nil {
+		f.reqHeaders = http.RequestHeader{RequestHeader: &fasthttp.RequestHeader{}}
+		h.SetRequestHeaderValue(ctx, name, value)
+	} else if req, ok := hdr.(http.RequestHeader); ok {
+		delHeader(req, name) // set doesn't currently overwrite all fields!
+		req.Set(name, value)
+	} else {
+		panic(fmt.Errorf("BUG: unknown type of request header: %v", h))
+	}
+}
+
+func (h host) AddRequestHeaderValue(ctx context.Context, name, value string) {
+	f := filterFromContext(ctx)
+	if hdr := f.reqHeaders; hdr == nil {
+		h.SetRequestHeaderValue(ctx, name, value)
+	} else if req, ok := hdr.(http.RequestHeader); ok {
+		req.Add(name, value)
+	} else {
+		panic(fmt.Errorf("BUG: unknown type of request header: %v", hdr))
+	}
+}
+
+func (host) RemoveRequestHeader(ctx context.Context, name string) {
+	f := filterFromContext(ctx)
+	if hdr := f.reqHeaders; hdr == nil {
+		// noop
+	} else if req, ok := hdr.(http.RequestHeader); ok {
+		delHeader(req, name)
+	} else {
+		panic(fmt.Errorf("BUG: unknown type of request header: %v", hdr))
+	}
+}
+
+func delHeader(h api.HeaderMap, name string) {
+	// TODO: httpwasm Del only deletes only a single value
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	h.Range(func(key, value string) bool {
+		if key == name {
+			h.Del(key)
+		}
+		return true
+	})
+}
+
+func (host) RequestBodyReader(ctx context.Context) io.ReadCloser {
+	b := filterFromContext(ctx).reqBody.Bytes()
+	return io.NopCloser(bytes.NewReader(b))
+}
+
+func (host) RequestBodyWriter(ctx context.Context) io.Writer {
+	f := filterFromContext(ctx)
+	f.reqBody.Reset()
+	return writerFunc(f.WriteRequestBody)
+}
+
+func (host) GetRequestTrailerNames(ctx context.Context) (names []string) {
+	return // no-op because trailers are unsupported: httpwasm/httpwasm#2145
+}
+
+func (host) GetRequestTrailerValues(ctx context.Context, name string) (values []string) {
+	return // no-op because trailers are unsupported: httpwasm/httpwasm#2145
+}
+
+func (host) AddRequestTrailerValue(ctx context.Context, name, value string) {
+	// panic because the user should know that trailers are not supported.
+	panic("trailers unsupported: httpwasm/httpwasm#2145")
+}
+
+func (host) RemoveRequestTrailer(ctx context.Context, name string) {
+	// panic because the user should know that trailers are not supported.
+	panic("trailers unsupported: httpwasm/httpwasm#2145")
+}
+
+func (host) SetRequestTrailerValue(ctx context.Context, name, value string) {
+	// panic because the user should know that trailers are not supported.
+	panic("trailers unsupported: httpwasm/httpwasm#2145")
+}
+
+func (host) GetURI(ctx context.Context) string {
+	// TODO(anuraaga): There is also variable.GetProtocolResource(ctx, api.URI), unclear if they must be kept in sync.
+	p := mustGetString(ctx, types.VarPath)
+	q, qErr := variable.GetString(ctx, types.VarQueryString)
+	if qErr != nil {
+		// No query, so an error is returned.
+		return p
+	}
+	return fmt.Sprintf("%s?%s", p, q)
```

Comment:
Should you urlencode `q`?

Reply:
yeah we should clarify that as raw I think in the abi doc, then update the base test with a value that is different if URL encoded. Wanna try this as a warm up change?

https://github.com/http-wasm/http-wasm/blob/main/content/http-handler-abi.md#set_uri
https://github.com/http-wasm/http-wasm-host-go/blob/main/handler/nethttp/host.go#L57-L76
https://github.com/http-wasm/http-wasm-host-go/blob/main/internal/test/testdata/e2e/uri.wat#L17
https://github.com/http-wasm/http-wasm-host-go/blob/main/handler/nethttp/middleware_test.go#L108-L146